File: javascript/decimal_to_roman_numbers.js
const symbolsWithValue = [
  { symbol: 'M', value: '1000' },
  { symbol: 'CM', value: '900' },
  { symbol: 'D', value: '500' },
  { symbol: 'CD', value: '400' },
  { symbol: 'C', value: '100' },
  { symbol: 'XC', value: '90' },
  { symbol: 'L', value: '50' },
  { symbol: 'XL', value: '40' },
  { symbol: 'X', value: '10' },
  { symbol: 'IX', value: '9' },
  { symbol: 'V', value: '5' },
  { symbol: 'IV', value: '4' },
  { symbol: 'I', value: '1' }
]

// Algorithm to resolve this problem
// 1. we need to divide the decimalNumber by symbolsWithName symbolsWithValue
// 2. Based on the quotient we are going to repeat that number of key values within the symbolsWithValue HashChangeEvent
// 3. We are going to calculate the modulus of the division operation and that number will be the new value for decimal_number
// 4. We repeat this process as many times as the decimal_number is zero

const convertDecimalToRomanValues = (decimalNumber) => {
  let finalRomanNumber = '';
  while (decimalNumber > 0) {
    symbolsWithValue.forEach((element) => {
      let division = Math.floor(decimalNumber / element.value);
      while (division--) {
        finalRomanNumber += element.symbol;
      }
      decimalNumber = decimalNumber % element.value;
    })
  }
  return finalRomanNumber;
};

console.log(`The roman number is ${convertDecimalToRomanValues(2999)}`);

File: ruby/decimal_to_roman_numbers.rb
@symbols_with_value = {
  'M': '1000',
  'CM': '900',
  'D': '500',
  'CD': '400',
  'C': '100',
  'XC': '90',
  'L': '50',
  'XL': '40',
  'X': '10',
  'IX': '9',
  'V': '5',
  'IV': '4',
  'I': '1'
}

# the time complexity is
# O(N)
def decimal_to_roman(decimal_number)
  @final_value = ''
  loop do
    @symbols_with_value.each do |symbol, decimal_value|
      division = (decimal_number / decimal_value.to_i).ceil

      division.times do
        @final_value += symbol.to_s
      end
      decimal_number = decimal_number % decimal_value.to_i
    end

    break if decimal_number <= 0
  end
  @final_value
end

puts decimal_to_roman(2999)
